### ISSUE TITLE
Question: data manipulation on event emission

### ISSUE THREAD

**Issue by Lautenschlager-id** [httptoolkit/mockttp#53]
Hello.

I've read the documentation thoroughly but I couldn't find a concrete answer. Reading the source didn't help much, either.

I'm currently mapping some requests with
```JS
server.on("request", function( request ) {
    let json = request.body.getJson();
    json.field = "????";
    //request.body.json = json;
});
```

But apparently there's no way for me to mock/edit the data from these fields.

I have tried several ways, even as the following (as buffer was apparently raw [not a copy])
```JS
let data = request.body.buffer.toString("utf8");
data = JSON.parse(data);
data.field = "???";
request.body.buffer = Buffer.from(JSON.stringify(data), "utf8");
```

But I failed to edit anything.

What I had to do is to set a `transformRequest` even before knowing whether the endpoint would be called or not, which is not the behavior I really planned to have.
```JS
server.post("path")
    .once() // I don't even know if the request is truly being called, but I was forced to use this
    .thenPassThrough({
        transformRequest:{
            updateJsonBody:{
                field: "???"
            }
        }
    });
    
server.on("request", (request) => {
    // field doesn't show "???" as I assume it should due to the chunk above,
    // but the request is sent with the mocked data correctly and I get the expected results (response)
    let field = request.body.getJson().field; 
})
```
---
TL;DR: Is it possible to edit the request (and/or response) data from the EventEmitter emission?

**Comment by pimterry** [comment]
Hi @Lautenschlager-id!

The short answer is: you can't do that with the `.on()` methods, but you can do the same thing in other ways.

The `.on()` methods can be useful for many cases, but purely for observing requests & responses as they're received from/sent to the client. They run async and totally independently from all request handling, so you can't transform anything there, and you'll always see the request and response from the perspective of the client.

Instead, you probably want something like this:

```javascript
server.post("/path")
    .thenPassThrough({
        beforeRequest: (req) => { // <-- beforeRequest, not transformRequest
            // This will be called for every matching request, as it's
            // received, before it's sent upstream. You can look at `req`
            // to make any decisions you like, or record anything about
            // the request itself.

            // To change the request, you can return an object with various
            // fields that define what should change, for example:
            return {
                method: 'PUT',
                json: { ...req.body.json, field: '???' }
            }
        }
    });
```

This will allow you to transform proxied requests as they're sent on upstream. The full documentation for the passthrough options is [here](https://httptoolkit.github.io/mockttp/interfaces/_rules_requests_request_handlers_.passthroughhandleroptions.html#beforerequest).

The `transformRequests` option that you used is very similar to this, but has a few differences: it must be a fixed transformation declared up front, as you've discovered, so only some specific transformations are supported, but in return it's generally simpler & clearer, and it's usually more performant than a custom callback, especially if you're using a remote client or intercepting streaming requests.

In the above code, I'm only matching POST requests to a specific path, but you could use `server.anyRequest().thenPassThrough(...)` to match all requests and implement completely custom logic for every request instead, if you'd like.

It's not totally clear what you're trying to do, but if you are actually looking to dynamically mock requests without sending them upstream at all, you could instead handle the request entirely within Mockttp by using [`.thenCallback((req) => { ... }))`](https://httptoolkit.github.io/mockttp/classes/_rules_requests_request_rule_builder_.requestrulebuilder.html#thencallback).

**Comment by Lautenschlager-id** [comment]
Hey! Thank you for the quick reply ;)

Oh, that's unfortunate to know. Then the request/response events are merely read-only. :(

What I currently need then is to mock the path `once`, but being capable of "removing" the listener if the endpoint won't be called. The thing is that there's a small chance that I won't get called, but it could be in other flows of my application. So, if it's not called in this specific flow, I would remove the listener because I don't want to mock any other screen!

For example:
```JS
function flow1(){
    let any_reference_to_cancel_this_listener = server.post("path").once().then.....

    let executed = waitForRequest("path") // pseudo
    if (!executed)
        server.cancelListener(any_reference_to_cancel_this_listener) // Is there such a thing? How do I cancel this `once`? 
}

function flow2(){
    // same path, i don't want the once_listener from flow1 to affect this call.
    // If it as called once, ok, it got killed. But if the endpoint in flow1 wasnt called, I don't want that mapping to mock flow2.
    let anything = waitForRequest("path")
}
```

**Comment by pimterry** [comment]
If you're using `beforeRequest` then you can use any logic you like at all within the rule, so one option is to include this logic on your mock. For example:

```javascript
let currentFlow = 'flow1';

server.post("/path")
    .thenPassThrough({
        beforeRequest: (req) => {
            if (currentFlow !== 'flow1') return;
            else return {
                json: { ...req.body.json, field: '???' }
            };
        }
    });
```

This will add `field` to JSON body responses for that endpoint, only whilst `currentFlow` is `flow1`. The rest of the time, the request will be proxied to the real server as normal unchanged. Does that work for your case?

Right now, you can't cancel an individual rule, so the above is the best option. Alternatively, you can reset all rules (either by stopping and starting the server, or calling `server.reset()`) so for many use cases you might want to do that instead. For example, you can start & stop or `reset()` Mockttp between each of your tests or steps, define a `setupMocks()` function that sets up any rules that are shared between all your tests or steps, and call that for each step plus adding any extra step-specific rules.

Would that work? It might help if you could tell me more about what you're trying to do otherwise, I think this might be clearer with some context. It could be possible to remove individual rules, but I don't think it should be necessary for most use cases that I'm aware of.

**Comment by Lautenschlager-id** [comment]
Oh, I see.

Well, `server.reset` will definitely solve the issue I'm facing right now :) But it would be great to be capable of deleting rule by rule, hehe.

Thank you for the support!
